diff --git a/archival/unzip60/win32/nt.c b/archival/unzip60/win32/nt.c
index a6cf8590..3432134c 100644
--- a/win32/nt.c
+++ b/win32/nt.c
@@ -37,8 +37,8 @@
 
 #define WIN32_LEAN_AND_MEAN
 #define UNZIP_INTERNAL
-#include "../unzip.h"
 #include <windows.h>
+#include "../unzip.h"
 #ifdef __RSXNT__
 #  include "../win32/rsxntwin.h"
 #endif
diff --git a/archival/unzip60/win32/win32.c b/archival/unzip60/win32/win32.c
index 410207f4..48ca4d5a 100644
--- a/win32/win32.c
+++ b/win32/win32.c
@@ -54,8 +54,8 @@
 
 
 #define UNZIP_INTERNAL
-#include "../unzip.h"
 #include <windows.h>    /* must be AFTER unzip.h to avoid struct G problems */
+#include "../unzip.h"
 #ifdef __RSXNT__
 #  include "../win32/rsxntwin.h"
 #endif
