diff --git a/src/readcdf.c b/src/readcdf.c
index 30c3d246..eeade572 100644
--- a/src/readcdf.c
+++ b/src/readcdf.c
@@ -236,7 +236,8 @@ cdf_file_property_info(struct magic_set *ms, const cdf_property_info_t *info,
 				} else {
 					char *c, *ec;
 					cdf_timestamp_to_timespec(&ts, tp);
-					c = cdf_ctime(&ts.tv_sec, tbuf);
+					time_t sec = ts.tv_sec;
+					c = cdf_ctime(&sec, tbuf);
 					if (c != NULL &&
 					    (ec = strchr(c, '\n')) != NULL)
 						*ec = '\0';
