diff --git a/configure.ac b/configure.ac
index a21d750..ee84810 100644
--- a/configure.ac
+++ b/configure.ac
@@ -228,7 +228,8 @@ if test "$os_unix" = "yes"; then
 	                          [Whether vsock support available])
 	                AC_MSG_RESULT([yes])],
 	               [AC_MSG_RESULT([no])])
-
+else
+    AM_CONDITIONAL([NEED_READPASSPHRASE], [false])
 fi
 
 # These are thngs we can work around
