diff --git a/src/assuan-support.c b/src/assuan-support.c
index c6052194..f36a759d 100644
--- a/src/assuan-support.c
+++ b/src/assuan-support.c
@@ -126,19 +126,19 @@ my_close (assuan_context_t ctx, assuan_fd_t fd)
 }
 
 
-static gpgme_ssize_t
+static ssize_t
 my_read (assuan_context_t ctx, assuan_fd_t fd, void *buffer, size_t size)
 {
   (void)ctx;
-  return _gpgme_io_read ((int) fd, buffer, size);
+  return (ssize_t) _gpgme_io_read ((int) fd, buffer, size);
 }
 
 
-static gpgme_ssize_t
+static ssize_t
 my_write (assuan_context_t ctx, assuan_fd_t fd, const void *buffer, size_t size)
 {
   (void)ctx;
-  return _gpgme_io_write ((int) fd, buffer, size);
+  return (ssize_t) _gpgme_io_write ((int) fd, buffer, size);
 }
 
 
@@ -181,7 +181,7 @@ my_sendmsg (assuan_context_t ctx, assuan_fd_t fd, const assuan_msghdr_t msg,
    to reflect the value of the FD in the peer process (on
    Windows).  */
 static int
-my_spawn (assuan_context_t ctx, pid_t *r_pid, const char *name,
+my_spawn (assuan_context_t ctx, assuan_pid_t *r_pid, const char *name,
 	  const char **argv,
 	  assuan_fd_t fd_in, assuan_fd_t fd_out,
 	  assuan_fd_t *fd_child_list,
@@ -288,7 +288,7 @@ my_spawn (assuan_context_t ctx, pid_t *r_pid, const char *name,
     {
       err = _gpgme_io_spawn (name, (char*const*)argv,
                              (IOSPAWN_FLAG_NOCLOSE | IOSPAWN_FLAG_DETACHED),
-                             fd_items, atfork, atforkvalue, r_pid);
+                             fd_items, atfork, atforkvalue, (pid_t*)r_pid);
     }
   if (!err)
     {
@@ -309,8 +309,8 @@ my_spawn (assuan_context_t ctx, pid_t *r_pid, const char *name,
 
 
 /* If action is 0, like waitpid.  If action is 1, just release the PID?  */
-static pid_t
-my_waitpid (assuan_context_t ctx, pid_t pid,
+static assuan_pid_t
+my_waitpid (assuan_context_t ctx, assuan_pid_t pid,
 	    int nowait, int *status, int options)
 {
   (void)ctx;
@@ -351,20 +351,20 @@ my_socketpair (assuan_context_t ctx, int namespace, int style,
 }
 
 
-static int
+static assuan_fd_t
 my_socket (assuan_context_t ctx, int namespace, int style, int protocol)
 {
   (void)ctx;
-  return _gpgme_io_socket (namespace, style, protocol);
+  return (assuan_fd_t) _gpgme_io_socket (namespace, style, protocol);
 }
 
 
 static int
-my_connect (assuan_context_t ctx, int sock, struct sockaddr *addr,
+my_connect (assuan_context_t ctx, assuan_fd_t sock, struct sockaddr *addr,
 	    socklen_t length)
 {
   (void)ctx;
-  return _gpgme_io_connect (sock, addr, length);
+  return _gpgme_io_connect ((int) sock, addr, length);
 }
 
 
diff --git a/src/gpgme-w32spawn.c b/src/gpgme-w32spawn.c
index 03ad1fbf..76befda6 100644
--- a/src/gpgme-w32spawn.c
+++ b/src/gpgme-w32spawn.c
@@ -249,7 +249,7 @@ my_spawn (wchar_t **argv, struct spawn_fd_item_s *fd_list, unsigned int flags)
           handle = LoadLibraryA ("user32.dll");
           if (handle)
             {
-              func = GetProcAddress (handle, "AllowSetForegroundWindow");
+              func = (BOOL (WINAPI *)(DWORD)) GetProcAddress (handle, "AllowSetForegroundWindow");
               if (!func)
                 FreeLibrary (handle);
             }
diff --git a/tests/gpg/t-encrypt-large.c b/tests/gpg/t-encrypt-large.c
index fd389de7..d4dd32b7 100644
--- a/tests/gpg/t-encrypt-large.c
+++ b/tests/gpg/t-encrypt-large.c
@@ -42,7 +42,7 @@ struct cb_parms
 
 
 /* The read callback used by GPGME to read data. */
-static ssize_t
+static gpgme_ssize_t
 read_cb (void *handle, void *buffer, size_t size)
 {
   struct cb_parms *parms = handle;
@@ -55,7 +55,7 @@ read_cb (void *handle, void *buffer, size_t size)
 }
 
 /* The write callback used by GPGME to write data. */
-static ssize_t
+static gpgme_ssize_t
 write_cb (void *handle, const void *buffer, size_t size)
 {
   struct cb_parms *parms = handle;
