diff --git a/lib/Makefile.am b/lib/Makefile.am
index 85f2b41..02544dd 100644
--- a/lib/Makefile.am
+++ b/lib/Makefile.am
@@ -68,14 +68,14 @@ gendata$(EXEEXT): gendata.c
 	$(CC) $(DEFAULT_INCLUDES) $(AM_CPPFLAGS) $(CFLAGS) $< -o $@
 data.c: $(srcdir)/gendata.c $(IDNA_TABLE)
 	$(MAKE) gendata$(EXEEXT)
-	$(builddir)/gendata$(EXEEXT) < $(IDNA_TABLE) > $@.new
+	$(EXE_WRAPPER) $(builddir)/gendata$(EXEEXT) < $(IDNA_TABLE) > $@.new
 	mv $@.new $@
 
 gentr46map$(EXEEXT): gentr46map.c
 	$(CC) $(DEFAULT_INCLUDES) $(AM_CPPFLAGS) $(CFLAGS) $< -o $@
 tr46map_data.c: $(srcdir)/gentr46map.c $(TR46MAP) $(NFCQC)
 	$(MAKE) gentr46map$(EXEEXT)
-	$(builddir)/gentr46map$(EXEEXT) > $@.new
+	$(EXE_WRAPPER) $(builddir)/gentr46map$(EXEEXT) > $@.new
 	mv $@.new $@
 
 $(IDNA_TABLE):
diff --git a/doc/Makefile.am b/doc/Makefile.am
index 720df52..1cb8b05 100644
--- a/doc/Makefile.am
+++ b/doc/Makefile.am
@@ -51,7 +51,7 @@ idn2.1: idn2.h2m $(top_srcdir)/src/idn2.c $(top_srcdir)/src/idn2.ggo \
 		--output=$@ $(top_builddir)/src/idn2$(EXEEXT)
 
 idn2-help.texi: $(top_srcdir)/src/idn2.c
-	$(top_builddir)/src/idn2$(EXEEXT) --help | $(GREP) '^  *-' > tmp
+	$(EXE_WRAPPER) $(top_builddir)/src/idn2$(EXEEXT) --help | $(GREP) '^  *-' > tmp
 	mv tmp idn2-help.texi
 
 MAINTAINERCLEANFILES = $(dist_man_MANS) decode.c example-tounicode.c example-toascii.c lookup.c \
